daml-script/daml3/Daml/Script/Questions/Submit.daml

```diff
@@ -10,168 +11,258 @@ module Daml.Script.Questions.Submit
+-- ```
+-- Party
+-- [Party]
+-- NonEmpty Party
+-- Set Party
+-- Optional Party
+-- ```
+actAs : IsParties p => p -> SubmitOptions
+actAs p = SubmitOptions (toParties p) [] []
+
+-- | Builds a SubmitOptions with given readAs parties.
+-- A given submission may omit any readAs parties and still be valid.
+-- Note that the parties type is constrainted by `IsParties`, allowing for specifying parties as any of the following:
+-- ```
+-- Party
+-- [Party]
+-- NonEmpty Party
+-- Set Party
+-- Optional Party
+-- ```
+readAs : IsParties p => p -> SubmitOptions
+readAs p = SubmitOptions [] (toParties p) []
+
+-- | Provides many Explicit Disclosures to the transaction.
+discloseMany : [Disclosure] -> SubmitOptions
+discloseMany ds = SubmitOptions [] [] ds
+
+-- | Provides an Explicit Disclosure to the transaction.
+disclose : Disclosure -> SubmitOptions
+disclose d = discloseMany [d]
+
+-- | Defines a type that can be transformed into a SubmitOptions
+class IsSubmitOptions a where
+  toSubmitOptions : a -> SubmitOptions
+
+-- Even though `IsParties SubmitOptions` doesn't exist, ghc doesn't consider context in instance selection.
+instance {-# OVERLAPS #-} IsSubmitOptions SubmitOptions where
+  toSubmitOptions = identity
+
+-- | Allows a single (or many) parties to be directly provided as actAs parties.
+-- This gives rise to the familiar syntax of
+-- ```
+-- alice `submit` createCmd MyContract with party = alice
+-- ```
+instance IsParties p => IsSubmitOptions p where
+  toSubmitOptions = actAs
+
+-- | Defines an applicative that can run transaction submissions. Usually this is simply `Script`.
+class Applicative m => ScriptSubmit m where
+  submitInternal : HasCallStack => SubmitOptions -> ErrorBehaviour -> Commands a -> m (Either (Optional SubmitError) (a, Optional TransactionTree))
+
+-- | Applicative that allows for multiple concurrent transaction submissions
+-- See `concurrently` for usage of this type.
+data ConcurrentSubmits a = ConcurrentSubmits with
+    submits : [Submission]
+    continue : [Either (Optional SubmitError) ([CommandResult], Optional TransactionTree)] -> a
+  deriving Functor
+
+instance Applicative ConcurrentSubmits where
+  pure : a -> ConcurrentSubmits a
+  pure x = ConcurrentSubmits [] (const x)
+
+  liftA2 : (a -> b -> c) -> ConcurrentSubmits a -> ConcurrentSubmits b -> ConcurrentSubmits c
+  liftA2 f a b = ConcurrentSubmits
+    { submits = a.submits <> b.submits
+    , continue = \results -> 
+        let (aResults, bResults) = splitAt (length a.submits) results
+        in f (a.continue aResults) (b.continue bResults)
+    }
+
+actAsNonEmpty : [Party] -> NonEmpty Party
+actAsNonEmpty ps = case ps of
+  [] -> error "actAs must be non-empty"
+  p :: ps -> NonEmpty p ps
+
+instance ScriptSubmit ConcurrentSubmits where
+  submitInternal : HasCallStack => SubmitOptions -> ErrorBehaviour -> Commands a -> ConcurrentSubmits (Either (Optional SubmitError) (a, Optional TransactionTree))
```

Comment:
I think we could move this implementation of `submitInternal` to the top level, then `class ScriptSubmit` could just be

```
class Applicative m => ScriptSubmit m where
    liftSubmission : HasCallStack => ConcurrentSubmits a -> m a
instance ScriptSubmit ConcurrentSubmits where
    liftSubmission : HasCallStack => ConcurrentSubmits a -> ConcurrentSubmits a
    liftSubmission = identity
instance ScriptSubmit Script where
    liftSubmission : HasCallStack => ConcurrentSubmits a -> Script a
    liftSubmission = concurrently
```

Reply:
Oh nice! Thats quite clean

Reply:
😄 thanks I'm glad you like it